src/Sql/dbo/Stored Procedures/CollectionUser_UpdateUsers.sql

```diff
@@ -14,35 +14,70 @@ BEGIN
             [Id] = @CollectionId
     )
 
-    ;WITH [AvailableUsersCTE] AS(
-        SELECT
-            Id
-        FROM
-            [dbo].[OrganizationUser]
-        WHERE
-            OrganizationId = @OrgId
+    CREATE TABLE #TempAvailableUsers
+    (
+        [Id] UNIQUEIDENTIFIER NOT NULL
     )
-    MERGE
+
+    INSERT INTO #TempAvailableUsers
+    SELECT
+        [Id]
+    FROM
+        [dbo].[OrganizationUser]
+    WHERE
+        [OrganizationId] = @OrgId
+
+    -- Update
+    UPDATE
+        [Target]
+    SET
+        [Target].[ReadOnly] = [Source].[ReadOnly]
+    FROM
         [dbo].[CollectionUser] AS [Target]
-    USING 
-        @Users AS [Source]
-    ON
+    INNER JOIN
+        @Users AS [Source] ON [Source].[Id] = [Target].[OrganizationUserId]
+    WHERE
         [Target].[CollectionId] = @CollectionId
-        AND [Target].[OrganizationUserId] = [Source].[Id]
-    WHEN NOT MATCHED BY TARGET
-    AND [Source].[Id] IN (SELECT [Id] FROM [AvailableUsersCTE]) THEN
-        INSERT VALUES
-        (
-            @CollectionId,
-            [Source].[Id],
-            [Source].[ReadOnly]
+        AND [Target].[ReadOnly] != [Source].[ReadOnly]
+
+    -- Insert
+    INSERT INTO
+        [dbo].[CollectionUser]
+    SELECT
+        @CollectionId,
+        [Source].[Id],
+        [Source].[ReadOnly]
+    FROM
+        @Users AS [Source]
+    WHERE
+        [Source].[Id] IN (SELECT [Id] FROM #TempAvailableUsers)
```

Comment:
Did this get you a better query plan than using an `INNER JOIN #TempAvailableUsers` against `@Users`?

Reply:
I don't think that I tried an inner join here, but I assume it would probably not be much different given the fact that `#TempAvailableUsers` is already filtered to a limited result set. Good idea though. Let me check it.

Reply:
I was able to get rid of the temp tables altogether with a join solution.